docs/snapshotting/design.md

```diff
@@ -0,0 +1,211 @@
+  - *The vmm snapshot file* used to load from, is released and no longer used by this process.
+- _on failure_: will report the encountered error, then end this Firecracker process (as it might be in an invalid state).
+
+**Return value:** Will report success if the microVM was successfully loaded, failure otherwise.
+**Body:**
+```yaml
+SnapshotLoadConfig:
+  type: object
+  required:
+    - snapshot_path
+    - mem_file_path
+  properties:
+    snapshot_path:
+      type: string
+      description: Path to file which contains the vmm state to be loaded.
+    mem_file_path:
+      type: string
+      description: Path to file which contains the guest memory to be loaded.
+                   This file should not be externally changed for the entire
+                   lifetime of the microVM to be loaded.
+```
+
+## Snapshot Format
+
+We will use **serde** as a serialization framework since it is a well maintained project with a large community. Serde defines a set of primitives and provides an interface for serializers to implement. Our approach is to build a custom serde serializer that meets the following criteria:
+
+* Performance: Serialization/deserialization should be fast and the snapshot small in size.
+* Code should be well-tested and documented.
+* Implementing cross-version snapshot restore should be easy to do.
+* The format should follow a clearly defined specification.
+* The format should have support in other languages, so tooling can be built around the snapshot format.
+* There should either be a large community that provides good support, or we should own it.
+
+### Current favorites
+
+1. Extend the existing serde compliant **bincode** serializer to match our requirements.
+2. Implement a **custom CBOR** serializer. It will require more work than any of the other approaches but this can be a very good option as we only keep serde as a dependency and there are no external risks factors for breaking changes in the snapshot format.
+
+### We also looked at these other options
+
+1. **Serde-bincode**. Use the upstream [`bincode`](https://crates.io/crates/bincode) crate and add minimal versioning support via using a snapshot translator.
+2. **Serde-cbor**. Use the upstream [`serde-cbor`](https://crates.io/crates/serde_cbor) crate and the [CBOR](https://tools.ietf.org/html/rfc7049) format.
+    It combines the strong type definition specific to JSON with a binary format for small snapshot sizes. Does not have out-of-the-box support for versioning. Early experimentation indicates potential need for upstream bugfixes.
+3. **Serde-protobuf**. We haven't invested much in researching protobuf, but it doesn't seem a good choice. It looks complicated to implement in Firecracker as the [`serde-protobuf` crate](https://crates.io/crates/serde-protobuf) is in its early stages. Another option is to directly use [`rust-protobuf`](https://github.com/stepancheg/rust-protobuf).
```

Comment:
The nice thing about protobuf is that it should be a lot more flexible than e.g. bincode. And definitely cross language, large community, etc.
But the serde crate for it is in its very early stages indeed.

Reply:
Protobuf is indeed an interesting choice, but using it forces us to create an external schema definition for each one of the structs that we want to serialize/deserialize. What we really want here is to decorate structures and use rust's derive macros to generate code.